### PR TITLE
Draft: Fix svg images not correctly displaying in e2ee rooms

### PR DESCRIPTION
**Pull request by justjanne**
Type: Defect
Related: https://github.com/vector-im/element-web/issues/15094
Related: https://github.com/vector-im/element-web/issues/18526

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Draft: Fix svg images not correctly displaying in e2ee rooms ([\#8325](https://github.com/matrix-org/matrix-react-sdk/pull/8325)). Contributed by @justjanne.

### CODE REVIEW

**Comment by justjanne** [comment]
This PR uses data URLs for embedded SVGs in E2EE rooms and other situations where no PNG version is available. This ensures we get a high quality preview without the risk of exposing our origin to user-defined scripts.

File: src/utils/MediaEventHelper.ts [hunk 1]
```diff
@@ -26,22 +26,46 @@ import { IDestroyable } from "./IDestroyable";
...
-    public readonly thumbnailBlob: LazyValue<Blob>;
+    private readonly sourceTypedBlob: LazyValue<ITypedBlob>;
+    private readonly thumbnailTypedBlob: LazyValue<ITypedBlob | null>;
+
+    public get sourceBlob() {
+        return this.sourceTypedBlob.value.then(it => it.data);
+    }
+    public get thumbnailBlob() {
+        return this.thumbnailTypedBlob.value.then(it => it.data);
+    }
```
Comment (turt2live) [inline src/utils/MediaEventHelper.ts:60]: these need return types, and a blank line between them

File: src/utils/MediaEventHelper.ts [hunk 2]
```diff
@@ -26,22 +26,46 @@ import { IDestroyable } from "./IDestroyable";
...
+}
+
 export class MediaEventHelper implements IDestroyable {
     // Either an HTTP or Object URL (when encrypted) to the media.
     public readonly sourceUrl: LazyValue<string>;
     public readonly thumbnailUrl: LazyValue<string>;
 
     // Either the raw or decrypted (when encrypted) contents of the file.
-    public readonly sourceBlob: LazyValue<Blob>;
-    public readonly thumbnailBlob: LazyValue<Blob>;
```
Comment (turt2live) [inline src/utils/MediaEventHelper.ts:36]: these were deliberately exposed and should not be hidden. It's for overall performance of the app.

File: src/utils/MediaEventHelper.ts [hunk 3]
```diff
@@ -59,46 +83,52 @@ export class MediaEventHelper implements IDestroyable {
...
         }
     };
 
-    private fetchSource = () => {
+    private fetchSource: () => Promise<ITypedBlob> = () => {
         if (this.media.isEncrypted) {
             const content = this.event.getContent<IMediaEventContent>();
-            return decryptFile(content.file, content.info);
+            return decryptFile(content.file, content.info).then(data => {
+                return { mimetype: content.info.mimetype, data };
```
Comment (turt2live) [inline src/utils/MediaEventHelper.ts:109]: `info` might not exist, and `decryptFile` should be responsible for mimetype detection.

File: src/utils/MediaEventHelper.ts [hunk 4]
```diff
@@ -59,46 +83,52 @@ export class MediaEventHelper implements IDestroyable {
...
-    private fetchThumbnail = () => {
+    private fetchThumbnail: () => Promise<ITypedBlob | null> = () => {
         if (!this.media.hasThumbnail) return Promise.resolve(null);
 
         if (this.media.isEncrypted) {
             const content = this.event.getContent<IMediaEventContent>();
             if (content.info?.thumbnail_file) {
-                return decryptFile(content.info.thumbnail_file, content.info.thumbnail_info);
+                return decryptFile(content.info.thumbnail_file, content.info.thumbnail_info).then(data => {
+                    return { mimetype: content.info.thumbnail_info.mimetype, data };
```
Comment (turt2live) [inline src/utils/MediaEventHelper.ts:122]: `info` might not exist, and `decryptFile` should be responsible for mimetype detection.

`thumbnail_info` also might not exist

File: src/utils/MediaEventHelper.ts [hunk 5]
```diff
@@ -26,22 +26,46 @@ import { IDestroyable } from "./IDestroyable";
...
+// infer type of blob from blob itself
+function toTypedBlob(blob: Blob): ITypedBlob {
+    return {
+        mimetype: blob.type,
+        data: blob,
+    };
+}
+
+async function createDataUrl(blob: ITypedBlob): Promise<string> {
+    return `data:${blob.mimetype};base64,${await blob.data.text().then(btoa)}`;
```
Comment (turt2live) [inline src/utils/MediaEventHelper.ts:43]: This opens us up to XSS attacks, which we do not want.

**Review by turt2live — CHANGES_REQUESTED** [review]
Thanks for taking a look at this, though it's not something we can merge easily. In addition to the comments below, I think this needs explicit review from the security team.

File: src/utils/MediaEventHelper.ts (hunk 2 above)
Comment (turt2live) [reply src/utils/MediaEventHelper.ts:36]: (also used by projects downstream of us)

**Comment by justjanne** [comment]
@turt2live I appreciate the review, after all this is just a first attempt at experimenting with potential solutions that allow us to get SVG previews working.

I'll experiment further with this over the easter weekend, though I don't expect that I'll get a clean fix done soon.

(I actually thought I had left the PR in draft status, but that may have gotten lost at some point)

**Comment by turt2live** [comment]
ah, apologies for the early review then 😅 

This is a relatively security sensitive area though, particularly the object URL handling.

**Pull request closed without merging**